### PR TITLE
rename Complex{32,64,128} to ComplexF{16,32,64}

### PR DESCRIPTION
**Pull request by rfourquet**
Cf. https://discourse.julialang.org/t/rename-complex2n-to-complexn/6081 for background.
Here is a recollection from this thread of 3 reasons to deprecate the `ComplexN` aliases:

- In `Complex128`, `128` refers to no semantic reality connected to the type. It only refers to the very low-level fact that it's its number of bits. Actually, in the initial discourse thread, I suggested renaming `Complex128` to `Complex64` with the idea "the coordinates of a `Complex64` is `Float64`", so there is a clearer connection between the 2 types.
- `Complex128` is ambiguous, since any eltype with 64 bit size (e.g. Int64) will give you a Complex type of size 128 (@Keno);
- Historically, the `Complex128` type actually preceded `Complex{Float64}` — it was originally a bits type because Julia didn’t have immutable types. So, at this point, deprecating `Complex128` makes a lot of sense (@stevengj).

### CODE REVIEW

File: base/linalg/blas.jl [hunk 1]
```diff
@@ -65,6 +65,9 @@ const liblapack = Base.liblapack_name
 
 import ..LinAlg: BlasReal, BlasComplex, BlasFloat, BlasInt, DimensionMismatch, checksquare, axpy!
 
+const Complex64 = Complex{Float32}
+const Complex128 = Complex{Float64}
```
Comment (ararslan) [inline base/linalg/blas.jl]: I'm confused. Isn't the point to remove definitions like this?

Comment (KristofferC) [reply base/linalg/blas.jl]: The BLAS module can define whatever convenience typealiases it wants though, right?

**Comment by alyst** [comment]
> I suggested renaming `Complex128` to `Complex64`

Is it the plan to deprecate the current `ComplexN` aliases and then to reintroduce them as `ComplexN/2`?

IIUC `Complex{Float64}` is by far the most commonly used complex type. It would be a pity not to have a shorter alias for it.

File: base/linalg/blas.jl (hunk 1 above)
Comment (ararslan) [reply base/linalg/blas.jl]: Well sure, but it seems like for clarity it might be nice to use the more explicit type names.

Comment (rfourquet) [reply base/linalg/blas.jl]: Yes, one reason I did that is that this file and in particular "lapack.jl" contain an awful lot of references to `Complex128/Complex64`, so I thougth it was fine to have local aliases (which I would certainly do for a personal project). Morever, I am really not familiar with the `LinAlg` module, so I prefer to receive suggestions from its authors as to what they prefer (alias or not). Also this module will eventually be split out from base, and so will these aliases.

**Comment by rfourquet** [comment]
> Is it the plan to deprecate the current ComplexN aliases and then to reintroduce them as ComplexN/2?

It would be my preference if there has to be aliases, but according to the discourse thread, it doesn't seem to considered as a good idea. But of course more opinion can be expressed and I really don't know at this point what will be decided.

File: base/sparse/umfpack.jl [hunk 2]
```diff
@@ -15,6 +15,8 @@ struct MatrixIllConditionedException <: Exception
     msg::AbstractString
 end
 
+const Complex128 = Complex{Float64}
```
Comment (Keno) [inline base/sparse/umfpack.jl]: Is this necessary?

File: base/linalg/lapack.jl [hunk 3]
```diff
@@ -14,6 +14,9 @@ import ..LinAlg: BlasFloat, Char, BlasInt, LAPACKException,
 
 using Base: iszero
 
+const Complex64 = Complex{Float32}
```
Comment (Keno) [inline base/linalg/lapack.jl]: Can we get rid of these as well?

**Review by Keno — APPROVED** [review]
LGTM. I'm fine either way with the local aliases in the linalg code. We could also teach femtocleaner to do this rewrite and have it be done automatically.

**Comment by rfourquet** [comment]
> could also teach femtocleaner to do this rewrite and have it be done automatically.

This could indeed be more safe. One of the other reason I used these local aliase is that `ComplexN` is used a lot as a `Symbol` which is interpolated, so the transformation would generally be e.g. `:Complex128` -> `Complex{Float64}`, which `sed` can do easily but this would need careful review. But again, I'm fine to do whatever linalg people tell me to do (if this PR is triage'd positively).

**Comment by KristofferC** [comment]
Consider using

```
julia> typeof(fill(0e0+0e0im, 5, 5))
Array{Complex{Float64},2}

julia> typeof(fill(0f0+0f0im, 5, 5))
Array{Complex{Float32},2}
```

or something of that sort.

**Comment by Sacha0** [comment]
Even simpler: `fill(0.0im, n, n)` :).

**Comment by JeffBezanson** [comment]
I'm not sure how I feel about this. I see that the names are a bit redundant and confusing, but it also seems to me that if the aliases are useful for the linalg code then they're useful for others as well.

**Comment by Sacha0** [comment]
> I see that the names are a bit redundant and confusing, but it also seems to me that if the aliases are useful for the linalg code then they're useful for others as well.

I favor removing those aliases from linalg and sparse. Best!

**Comment by rfourquet** [comment]
> if the aliases are useful for the linalg code then they're useful for others as well.

From what i could see, those types are used in linalg mainly for wrapping libraries, with loops and `eval`, so this is mainly to be exhaustive, but using `Complex{Float64}` instead of `Complex128` would'n hurt clarity I think, as these appeary only as variables of those loops. So actually I would favor also getting rid entirely of the aliases even there. I will do that tomorrow as there seem to be support for this.

**Comment by andyferris** [comment]
Other languages like maybe Go have these type names also, no? (Mostly, I would avoid the rename to `ComplexN/2` since it seems nonstandard across languages; I don’t tend to use `Complex128` myself).

**Comment by rfourquet** [comment]
Added the "decision" label. There is support here and on the discourse thread, but I'm not the one to make the final call. As this is breaking, I will resort to "triage" eventually if needed.

**Comment by StefanKarpinski** [comment]
No need to be that shy with the triage label, that's what it's for. We may as well discuss it.

**Comment by StefanKarpinski** [comment]
[NumPy](https://docs.scipy.org/doc/numpy-1.13.0/user/basics.types.html) and [Go](https://golang.org/pkg/builtin/#complex128) use analogous names. [Rust](http://huonw.github.io/primal/num_complex/type.Complex64.html) uses `Complex64` to mean `Complex{Float64}`.

**Comment by StefanKarpinski** [comment]
I'm fine with deprecating these types. We may want to introduce `ComplexF64` and `ComplexF128` at some point as shorthands for `Complex{Float64}` and `Complex{Float128}`, but that can be done any any point.

**Comment by garrison** [comment]
I've been thinking more about why I am opposed to this.  My main objection is about `Complex128`; the others are far less useful to me.  At the moment, I can write `zeros(N)` to get a `Vector` of `Float64`s, but if I want a vector of equivalent complex numbers I write `zeros(Complex128, N)`.  With this change, complex numbers feel even more like second class citizens, as the expression becomes `zeros(Complex{Float64}, N)`.

I think it is telling that the original pull request left in these aliases for use in the linear algebra code.  These aliases are useful and will be even more so in user code, especially for those of us using complex numbers frequently.

One thing I especially like about the status quo -- and this gets to the heart of my thumbs down -- is that the aliases are standardized.  I would hate to see various code bases define these aliases on their own.  Right now, if I see `Complex64` in julia code, I know precisely what it means; I do not need to wonder whether the local code base defined it to be `Complex{Float64}` or `Complex{Float32}`.  I think @StefanKarpinski's suggested `ComplexF...` names can help with this in the long run, if people can all agree on the name.  If the path forward is indeed to have aliases that are more clearly named, I would remove my objection completely.

As for ambiguity, it is already possible to learn what `Complex128` means by simply typing it into the repl:

```
julia> Complex128
Complex{Float64}
```

Best! [channeling @Sacha0 here... :wink:]

**Comment by Sacha0** [comment]
Regarding `zeros(Complex128, N)`, why not `fill(0.0im, N)`? The latter is both shorter and less ambiguous :).

**Comment by rfourquet** [comment]
> If the path forward is indeed to have aliases that are more clearly named, I would remove my objection completely.

I agree that it could be better to still have "official" aliases. I'm not a big user of `Complex` types myself, so I always have a (very) small moment of confusion when I deal with them, in that the translation from the `128` suffix to `Float64` is not automatic yet in my brain. In my original post on discourse (linked to in the OP), I was suggesting to rename `Complex128` to `Complex64`, which turned out to be a bad idea, but `ComplexF64` would be fine for me; and I was arguing that it's not scalable to name a type after it's total number of bits: I would start to get really confused with `Quaternion256` or `Octonion512` ;-p

So I understand that the removal of `triage` is a green light to go on with this, so I will rebase withing the next few days, and I volunteer to introduce the `ComplexF64` aliases if it's decided upon.

**Comment by garrison** [comment]
> why not `fill(0.0im, N)`?

On first glance this looks to me like initializing a purely imaginary number rather than a complex number.  Of course, julia does not have purely imaginary numbers, so it indeed results in a complex number after an additional mental step.  I believe I could get accustomed to this for initializing zeroed arrays.

Aside from initializing arrays, I often type `Complex128` in places where I suspect inference may not succeed, such as to specify the type of a comprehension explicitly.  My understanding is that inference is not meant to be predictable in general, but I do not have a sophisticated understanding of when (if ever) I should expect it to be predictable, which results in me specifying the type often.  I have run into numerous instances where type inference fails in places where I would expect it to be obvious by inspection (thus slowing my code down significantly).  For instance, the code I reported at https://github.com/JuliaLang/julia/issues/17717#issuecomment-236355708 behaves in recent versions as I would expect, but I can come up with slightly more complicated examples where inference still unexpectedly fails with comprehensions.  I am not sure if this is inherent or not (helping me understand this, of course, is a topic better suited for Discourse :wink:), but if it is indeed the case that many simple type-inference cases should not be predictable, then I believe the aliases are a lot more useful than they might otherwise be.

Best!

**Comment by Sacha0** [comment]
> On first glance this looks to me like initializing a purely imaginary number rather than a complex number. Of course, julia does not have purely imaginary numbers, so it indeed results in a complex number after an additional mental step. I believe I could get accustomed to this for initializing zeroed arrays.

Perhaps `fill(0.0 + 0.0im, N)` if you prefer? :)

**Comment by garrison** [comment]
Or `fill(complex(0.0), N)`, which is one fewer keystroke.  (`complex(0.)` also works, but I cannot recall whether there are plans to deprecate this.)  In either case, it is indeed as simple to type as `zeros(Complex128, N)`.

**Comment by stevengj** [comment]
`randn(Complex128, n)` seems like a better example than `zeros` here.   But I agree that these types are so commonplace that people will be tempted to make aliases, and we might as well keep "standard" aliases in Base.

**Comment by PallHaraldsson** [comment]
Since you're thinking this over, I was thinking if `ComplexPolar{Float64,Fixnum8}` might be valuable.

Would you like to see it then as ComplexPolar72?

**Comment by StefanKarpinski** [comment]
Please don't derail the discussion, @PallHaraldsson.

**Comment by rfourquet** [comment]
Triage decided to introduce new aliases, namely `ComplexF64` instead of `Complex128` etc.
I updated the PR to reflect that, by applying `sed` mainly (except "NEWS.md" and "deprecated.jl"), so it's less likely to contain human mistakes, but also it's possible one instance or two have been replaced when they shouldn't!

File: stdlib/SuiteSparse/src/cholmod_h.jl [hunk 4]
```diff
@@ -26,8 +26,8 @@ const COMPLEX = Int32(2)       # a complex matrix (ANSI C99 compatible)
 const ZOMPLEX = Int32(3)       # a complex matrix (MATLAB compatible)
 xtyp(::Type{Float32})    = REAL
 xtyp(::Type{Float64})    = REAL
-xtyp(::Type{Complex64})  = COMPLEX
-xtyp(::Type{Complex128}) = COMPLEX
+xtyp(::Type{ComplexF32})  = COMPLEX
+xtyp(::Type{ComplexF64}) = COMPLEX
```
Comment (rfourquet) [inline stdlib/SuiteSparse/src/cholmod_h.jl]: update alignment

File: base/complex.jl [hunk 5]
```diff
@@ -28,9 +28,9 @@ julia> im * im
 """
 const im = Complex(false, true)
 
-const Complex128 = Complex{Float64}
-const Complex64  = Complex{Float32}
-const Complex32  = Complex{Float16}
+const ComplexF64 = Complex{Float64}
+const ComplexF32  = Complex{Float32}
+const ComplexF16  = Complex{Float16}
```
Comment (rfourquet) [inline base/complex.jl]: update alignment

File: base/linalg/lapack.jl [hunk 6]
```diff
@@ -5356,8 +5356,8 @@ for (gecon, elty) in
 end
 
 for (gecon, elty, relty) in
-    ((:zgecon_,:Complex128,:Float64),
-     (:cgecon_,:Complex64, :Float32))
+    ((:zgecon_,:ComplexF64,:Float64),
+     (:cgecon_,:ComplexF32, :Float32))
```
Comment (rfourquet) [inline base/linalg/lapack.jl]: update alignment

File: stdlib/IterativeEigenSolvers/src/arpack.jl [hunk 7]
```diff
@@ -255,8 +255,8 @@ for (T, saupd_name, seupd_name, naupd_name, neupd_name) in
 end
 
 for (T, TR, naupd_name, neupd_name) in
-    ((:Complex128, :Float64, :znaupd_, :zneupd_),
-     (:Complex64,  :Float32, :cnaupd_, :cneupd_))
+    ((:ComplexF64, :Float64, :znaupd_, :zneupd_),
+     (:ComplexF32,  :Float32, :cnaupd_, :cneupd_))
```
Comment (rfourquet) [inline stdlib/IterativeEigenSolvers/src/arpack.jl]: update alignment

**Comment by antoine-levitt** [comment]
Not sure if triage closes the discussion definitely, but FWIW I don't like the added F. It's ugly, and its only advantage is to clarify that it's a floating point and not an integer type. Is that really a serious confusion to anyone? (there's https://en.wikipedia.org/wiki/Gaussian_integer, but that seems pretty esoteric...). Or is the intention to use ComplexF to ease deprecation, then switch to Complex again?

**Comment by rfourquet** [comment]
> Or is the intention to use ComplexF to ease deprecation, then switch to Complex again?

I don't think there is such a plan.

**Comment by StefanKarpinski** [comment]
There was fairly broad support that using the bitsize of the entire type in `ComplexN` was confusing to many people, and those who didn't mind it were only ambivalent. There's an unfortunate lack of consensus of what names like "complex64" means in different languages – in NumPy and Go it means `Complex{Float32}` whereas in Rust it means `Complex{Float64}`. The new names are unambiguous: the `F64` in `ComplexF64` can only refer to `Float64` and therefore must mean `Complex{Float64}`. Note that `ComplexF64` and `Complex128` are the same length, and while `ComplexF32` is one character longer than `Complex64` the lack of ambiguity seems worth it.

**Comment by rfourquet** [comment]
I will rebase and merge by tomorrow if no-one objects.

**Comment by garrison** [comment]
> in NumPy and Go [`Complex64`] means `Complex{Float32}`

[Also in Mathematica](https://reference.wolfram.com/language/ref/format/Complex128.html).  As far as I can tell, Rust is the odd one out.

**Comment by stevengj** [comment]
In Fortran `complex*2n` is likewise used for complex numbers made of `real*n`, although in that case n is in bytes. 

That being said, I guess I’m fine with the proposed new names too.

**Comment by StefanKarpinski** [comment]
Let's let this simmer for a bit longer.

**Comment by rfourquet** [comment]
If subjectivity could be an argument, I felt a nice stress relief when renaming `Complex128` to `ComplexF64` :stuck_out_tongue_winking_eye: 
Joke aside, I wonder how generic the languages having chosen `Complex128` are... And even if `Complex{Int}` seems to be niche, it will be used. What I really like about `Float64` is that it sets a precedent in naming conventions: the package implementing quaternions will have `QuaternionF64` be unambiguous, but also `PlanePointF64` or `SpacePointI32` will be easily interpreted etc.

**Comment by StefanKarpinski** [comment]
Well, I'm convinced 😀

**Comment by garrison** [comment]
`CFloat64` is an alternative that is more generalizable in some ways, and there are at least a few languages using this name.  But I think it is a complete non-starter for julia, due to potential confusion with the existing `Cint`, `Cdouble`, etc. types.

**Comment by StefanKarpinski** [comment]
`CFloat64` would definitely be confusing with our `Cint` etc. convention. I also think that the most salient part of the `Complex{Float64}` type is that it's complex, not that it's based on Float64.

**Comment by garrison** [comment]
> also `PlanePointF64` or `SpacePointI32`

This is outside the scope of the current pull request, but because the proposed naming convention suggests other aliases as well, how do people feel about the following?

```julia
const ComplexI8 = Complex{Int8}
const ComplexI16 = Complex{Int16}
const ComplexI32 = Complex{Int32}
const ComplexI64 = Complex{Int64}
const ComplexI128 = Complex{Int128}
const ComplexI = Complex{Int}
```

**Comment by garrison** [comment]
As for myself, I've removed my downvote of this PR, as my objection was to the original plan of removing the aliases entirely.  (At this point, my remaining bikeshedding is really about making sure we have the best possible aliases.)

**Comment by rfourquet** [comment]
Freshly rebased, looking for someone to do the honors!

**Comment by musm** [comment]
Personally, I don't feel like the code churn is really worth the renaming from `Complex{N}` to `ComplexF{N}`,   even so considering that the `F` postfix  in `ComplexF` isn't exactly standard and could be confusing to beginners in numerical computing.

**Comment by KristofferC** [comment]
> Personally, I don't feel like the code churn is really worth the renaming from Complex{N} to ComplexF{N}

That's not what this PR does though?

**Comment by Keno** [comment]
🎉

**Pull request merged**